lib/bookstore-scrapper/helpers/getTextbookInformation.ts

```diff
@@ -0,0 +1,180 @@
+const endParsePrice = "new";
+const quoteParsePrice = "'";
+const notAvailable = null;
+
+export interface CampusStoreEntry {
+  book: Book;
+  course: Course;
+  dept: Dept;
+}
+
+export const getTextbookInformation = async (
+  link: string,
+  program: string,
+  term: string,
+  dept: string,
+  course: string
+): Promise<CampusStoreEntry[]> => {
+  // Array to return
+  const campusStoreEntries: CampusStoreEntry[] = [];
+
+  // Get HTML source code of page
+  let bookHTML: string = (await axiosRequest(link)) as string;
+
+  // Check if page has additonal recommended supplies
+  const suppliesIndex: number = bookHTML.indexOf(parseEndBookHTML);
+  if (suppliesIndex > 0) bookHTML = bookHTML.substring(0, suppliesIndex);
+
+  // Split the different books on the page into parts of an array
+  let bookSplit = bookHTML.split(parseISBN);
+
+  // Find out the department
+  const deptSplit = dept.split("-");
+  let deptName = "";
+  const deptAbv = deptSplit[0].trim();
+  if (deptSplit.length === 1) {
+    deptName = deptAbv;
+  } else {
+    deptName = deptSplit[1].trim();
+  }
+  const deptData: Dept = {
+    id: randomUUID(),
+    name: deptName,
+    abbreviation: deptAbv,
+  };
+
+  // Find out the course
+  const courseName = await getCourseName(deptAbv, course);
+  const courseData: Course = {
+    id: randomUUID(),
+    name: courseName,
+    code: course,
+    term: term,
+    deptId: deptData.id,
+  };
+
+  // Create the book data object for each book
+  for (let book = 1; book < bookSplit.length; book++) {
+    const isbn = bookSplit[book].substring(0, 13);
+
+    const campusStorePrice = getBookPrice(bookHTML, isbn);
+
+    if (JSON.stringify(campusStoreEntries).includes(isbn)) {
+      continue;
+    }
+
+    let name = bookSplit[book].substring(
+      bookSplit[book].indexOf(parseBookName) + 5,
+      bookSplit[book].length
+    );
+    name = name.substring(0, name.indexOf(endParseSlash));
+
+    const imageUrl = `${bookImagePath}${isbn}`;
+
+    const googleBooksId = await getGoogleBooksId(isbn);
+
+    const bookData: Book = {
+      id: randomUUID(),
+      isbn,
+      name,
+      imageUrl,
+      googleBooksId,
+      isCampusStoreBook: true,
+      campusStorePrice,
+    };
+
+    const campusStoreEntry: CampusStoreEntry = {
+      book: bookData,
+      course: courseData,
+      dept: deptData,
+    };
+
+    campusStoreEntries.push(campusStoreEntry);
+  }
+  return campusStoreEntries;
+};
+
+const getCourseNameHTML = async (): Promise<string> => {
+  let page: number = 1;
+  let courseNameHTML: string = "";
+  while (!courseNameHTML.includes(finalPageText)) {
+    const courseNamePageHTML = (await axiosRequest(
+      `${startCourseNamePath}${page}${endCourseNamePath}`
+    )) as string;
+    courseNameHTML = courseNameHTML.concat(courseNamePageHTML);
+    page += 1;
+  }
+  return courseNameHTML;
+};
+
+const getCourseName = async (dept: string, code: string) => {
+  const courseNameHTML = await getCourseNameHTML();
```

Comment:
Moved this inside `getCourseName` to avoid top-level await. Can you double-check to make sure the logic is still good.

Reply:
`courseNameHTML` needs to be defined outside the function. The point for this is to make API calls that get the HTML containing the names for all the courses before parsing the campus store website.

Afterwards, `getCourseName` uses the same string defined in courseNameHTML to parse all course names. The reason of getting all the course names at the start (total of 33 network calls to pages containing names) is that the academic calendar website has some kind of throttling that I found out about while testing.

Making 33 network calls would be better than making over 2000 calls to the website while parsing through the campus store website pages. Is there something else that can be done to avoid this?

Reply:
Yeah, just needs to be refactored so that this function is only called as often as you need it (by another function)

Reply:
I see, okay. Moved it into another function and resolved this. Thanks Caleb.